### PR TITLE
Leap 48in24 approaches

### LINKED ISSUE #1400: Leap approaches for 48in24
As explained on [the forum](https://forum.exercism.org/t/48in24-exercise-01-16-leap/8964) we need approaches for the _leap_ exercise in as many languages as possible.

### PR DESCRIPTION
**Pull request by michalporeba**
This resolves #1400

I propose an introduction to the approaches and specific approaches to the problem. 
Considering the time until 48in24 I welcome early feedback on the text and the selection of the approaches. 
I will work on specific approaches in subsequent commits. 

The difference between `rem/2` and `Integer.mod/2` might be an interesting article to enrich the deep dive, however, due to time constraints and the need to write up approaches in other languages, I consider it out of scope for this PR.

### CODE REVIEW

File: exercises/practice/leap/.approaches/case/snippet.txt
Comment (Tuxified) [inline exercises/practice/leap/.approaches/case/snippet.txt:6]: Since this branch represents "all else", I feel a more idiomatic way to write it would be `_ -> false` since the shape of the tuple is not important anymore.

```suggestion
        _ -> false
```

**Comment by Tuxified** [comment]
This looks great, I'm not sure if explaining the difference between Kernel.rem/2 and Integer.mod/2 will be that interesting for the purpose of this challenge. Besides these approaches we could also add an if expression (translated from a Go solution):

```
def leap?(year) do
 if rem(year, 100) == 0 do
    rem(year, 400) == 0
  else
    rem(year, 4) == 0
  end
end
```

Really super small pet peeve: Elixir is usually indented using 2 spaces, not 4 ;)

**Review by Tuxified — APPROVED** [review]

**Comment by angelikatyborska** [comment]
I was holding off any reviews on this as it's still marked a draft, but let me just address:

> Really super small pet peeve: Elixir is usually indented using 2 spaces, not 4 ;)

I'm unhappy that the snippets have to be `.txt` files (as I learned on Discord the configuration option for the file format doesn't work, not your fault...) Before approving this PR, I intend to write some small script that will turn them to `.ex`, autoformat, and then turn back.

**Comment by angelikatyborska** [comment]
@michalporeba I pushed a commit with a script `bin/format_approach_snippets.sh` that you can run to format the code snippet. I couldn't find anything that could also format Elixir code in markdown files, so we will need to make sure manually that they are formatted correctly too.

**Comment by michalporeba** [comment]
Thanks for the feedback. Please don't wait. There is little time and there is still a lot to do. This is why I created a draft PR to ask for early feedback. I still have specific approaches to write up, and the snippets may change slightly once the approaches are written up. 

At the moment, I'm mostly interested in what you think about the structure of the introduction selections of the approaches. I'll start on the individual approaches tomorrow.

File: exercises/practice/leap/.approaches/cond/snippet.txt
Comment (michalporeba) [reply exercises/practice/leap/.approaches/cond/snippet.txt:6]: The function `divides?` is defined in the introduction document and is shared between approaches. Because of that, I don't think repeating it in individual snippets will be of value. The snippets, as I understand, should illustrate the approach, and they don't have to compile as such. It's just a code fragment. The reason I'm using the `divedes?` function is so that `rem(year, 100) == 0)` doesn't distract from what really differs.

File: exercises/practice/leap/.approaches/case/snippet.txt
Comment (michalporeba) [reply exercises/practice/leap/.approaches/case/snippet.txt:6]: It's something to discuss in the details of the approach, I think. I hope we will be able to discuss this once I have written up the approach.

**Comment by michalporeba** [comment]
> @michalporeba I pushed a commit with a script `bin/format_approach_snippets.sh` that you can run to format the code snippet. I couldn't find anything that could also format Elixir code in markdown files, so we will need to make sure manually that they are formatted correctly too.

Hi @angelikatyborska. The scrip appears to be causing [issues with checks](https://github.com/exercism/elixir/actions/runs/7452830847/job/20276931411?pr=1401). Considering that there will be more code in the .md files in the discussion than there will be in the snippets, perhaps we need to depend on manual code formatting? 

Snippets also often contain partial code, that's my experience from other tracks. The idea is to illustrate briefly what will be discussed on the approach page.  Won't that be a problem for the formatted?

**Comment by angelikatyborska** [comment]
That's on purpose, I want CI checks to fail if the code in snippets is not formatted. This will also require that the snippets are formally valid Elixir code, but it doesn't need to actually compile.

It would be nice if I could do the same for the code in `.md` files, but for now we need to manually format there.

File: exercises/practice/leap/.approaches/case/snippet.txt
Comment (angelikatyborska) [reply exercises/practice/leap/.approaches/case/snippet.txt:6]: Still, this snippet should use the more idiomatic version, which is the version @Tuxified suggested

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/introduction.md:44]: This sounds wrong somehow, but I can't quite define it... "we can (...)" sounds like it's optional? But it's not optional to put the clauses in the right order, the order always matters. Maybe we can remove this sentence from here
```suggestion
```

and then in the detailed approach description mention something like:
````
Remember that the order of function clauses matters, and only the first clause that matched a specific function call will get executed. All the other clauses will be ignored. This means that we **don't** need to repeat out checks:
```elixir
# Don't copy this!
def leap_year?(year) when rem(year, 400) == 0, do: true
def leap_year?(year) when rem(year, 100) == 0 and not rem(year, 400) == 0, do: false
def leap_year?(year) when rem(year, 4) == 0 and not rem(year, 100) == 0 and not rem(year, 400) == 0, do: true
def leap_year?(year) when not rem(year, 4) == 0 and not rem(year, 100) == 0 and not rem(year, 400) == 0,, do: false
```
````

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/introduction.md:57]: "the order of the checks can be done..." sounds wrong somehow. What about:
```suggestion
Similarly to the multiple clause function approach, we can also use a `cond` expression. 
```

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/introduction.md:41]: This approach should be called 
```suggestion
## Approach: Multiple clause function
```

Guards can be used either in a `case` expression or in multiple clause functions.

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/introduction.md:14]: Can you update all headings to only have the first word capitalized instead of all? That the style of headings we use in this repo
```suggestion
## General solution
```

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/introduction.md:70]: This approach should be called
```suggestion
## Approach: Using case on a tuple
```

There are different solutions that could be called "using case", for example any `cond` can be rewritten to a `case` (but this one is not very good, so let's not describe it 😉)

```elixir
case year do
  year when rem(year, 400) == 0 -> true
  year when rem(year, 100) == 0 -> false
  year when rem(year, 4) == 0 -> true
  _ -> false
end
```

**Review by angelikatyborska — COMMENTED** [review]
Thanks for working on this! I personally don't have the motivation to write any approaches myself, but if do, I promise to review all your PRs on time 😇.

For the future it's better to tell me explicitly that you want a draft PR reviewed when you open it. I usually wait until it's marked as "ready for review" because, well... that's the name of the feature 😂 and some people are shy about having their work in progress judged.

**Comment by angelikatyborska** [comment]
Over all, I think the approaches you've identified so far are the right ones to describe. I browsed some of the community solutions and most of them qualify as one of those approaches or a mix of two.

I found one extra approach that I would call "I love nesting", but I wouldn't recommend it to anyone 😁 
```elixir
def leap_year?(year) do
    if rem(year, 4) == 0 do
      if rem(year, 100) == 0 do
        if rem(year, 400) == 0 do
          true
        else
          false
        end
      else
        true
      end
    else
      false
    end
  end
```

File: exercises/practice/leap/.approaches/introduction.md
Comment (michalporeba) [reply exercises/practice/leap/.approaches/introduction.md:12]: I used the "non-negative" here very intentionally. Positives start at 1, non-negative start at 0, so the zero is included. Probably a better name for the `n` parameter and for the description would be to say 'divisor' as this is technically what it is, and where the behaviour of the two functions differ.

**Comment by michalporeba** [comment]
> Thanks for working on this! I personally don't have the motivation to write any approaches myself, but if do, I promise to review all your PRs on time 😇.

Thanks. I think I'm the opposite. I like the approaches and writing about ways to solve problems with code. I don't know for how long I will have enough energy to do so, but I'll approach this problem a week at the time. My objective is to finish this by the end of the week with all the detailed approaches. I know there is no reverse exercise for the second week, so it will give us (me) some time to prepare approaches for the following week. 

> 
> For the future, it's better to tell me explicitly that you want a draft PR reviewed when you open it. I usually wait until it's marked as "ready for review" because, well... that's the name of the feature 😂 and some people are shy about having their work in progress judged.

I have requested this on the forum, but I will make it more explicit in the PRs, too. Generally, here I'm also the opposite. "ready for review" for me is "ready for the final review and merge", while the "draft" is for early feedback that can lead to changes in direction. I'm not shy in that regard. I love early feedback and for my work in progress to be judged. 

Many comments you provided today will be very helpful as I write more text in the approaches to reduce, hopefully, the number of corrections required. 

Dzięki!

File: exercises/practice/leap/.approaches/introduction.md
Comment (michalporeba) [reply exercises/practice/leap/.approaches/introduction.md:20]: I was inspired by my solution to the problem I submitted in the Clojure track. I agree that many of the example or community solutions don't represent anything close to ideal. Sometimes, I see things starred the most, which I would consider an anti-pattern rather than a solution worth showing. 

I think that the approaches are a perfect place where the solutions can be curated to things worth showing rather than cataloguing what's possible. 

As to the divdes? The gain is not big for experienced developers, but for those less experienced, I think it might help with showing intent. When it comes to naming things I'm really inspired by [Kevlin Henny's Seven Ineffective Coding Habits of Many Programmers talk](https://www.youtube.com/watch?v=SUIUZ09mnwM). For me, the code is primarily an intent communication tool. 

But I'll think a bit more about it as I write out the details of the approaches.

**Comment by michalporeba** [comment]
@angelikatyborska, please have a look at [operators/content.md](https://github.com/exercism/elixir/pull/1401/files#diff-faee1c013e2b26e200ac02db5255eca19da2ec270977cf126ca9c77f05eae80f) file. This is the first full approach I have written up for this track, so I'd like some early feedback before I write up the remaining three. 

What do you think about the level of detail? Too little? Too much? How about external links? 

Some of the formatting is difficult to see and evaluate unless the deep dive page is visible on Exercism. That's something I have learnt in other tracks. I think we should assume that despite doing our best to get it right, there might need to be a subsequent PR with a few corrections (e.g., headers, spacing around code examples, snippet formatting).

File: exercises/practice/leap/.approaches/introduction.md
Comment (michalporeba) [reply exercises/practice/leap/.approaches/introduction.md]: I plan to show multiple options, similar to what I've done in the boolean operators approach. the `on a tuple` will be just one variant.

**Comment by michalporeba** [comment]
@angelikatyborska, @Tuxified, should it be `multiple clause functions` or `multiple function clauses`? The [documentation](https://hexdocs.pm/elixir/main/patterns-and-guards.html#guards) refers to `function clauses` rather than `clause functions`. However, the relevant [exercism page](https://exercism.org/tracks/elixir/concepts/multiple-clause-functions) has `multiple cause functions`.

**Comment by angelikatyborska** [comment]
> should it be `multiple clause functions` or `multiple function clauses`?

What's the context?

Here's a quick reference to how both phrases can be used correctly:

Screenshot 2024-01-13 at 08 45 38

File: exercises/practice/leap/.approaches/operators/content.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/operators/content.md:17]: I'd say short-circuiting does not matter at all in this exercise. Short-circuiting is important to consider when we want to avoid or ensure some side effects, for example:

```
iex()> 40 / x
** (ArithmeticError) bad argument in arithmetic expression: 40 / 0
    :erlang./(40, 0)
    iex:3: (file)

iex()> x != 0 && 40 / x
false
```

In this exercise, all inputs are valid, all the checks on either side of the boolean operator are pure functions with no side effects.

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [reply exercises/practice/leap/.approaches/introduction.md]: Hmm... I'm not entirely sure what the philosophy of writing Exercism approaches is supposed to be because I don't really read them. Are they supposed to cover every possible way of solving an exercise, or just the most common or interesting ones?

Because if we only want the common ones, then "case on a tuple" is the only case-based approach that I would suggest showing, based on what the community solutions tend to be. The rest of the case approaches that I see in the current version of the `exercises/practice/leap/.approaches/flow/content.md` file are less appropriate uses of `case` that should have been an `if` or a `cond` instead (like you point out already).

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/introduction.md:35]: Some typos
```suggestion
In the [boolean operators approach][operators-approach] we discuss the details of the solution.
It includes variations of the operators and their precedence.
```

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/introduction.md:48]: There is only one function in this exercise, `leap_year?`:
```suggestion
In the [multiple clause function approach][functions-approach] we discuss why in this approach the `Integer.mod/2` function will not work. 
```

Also, the directory name of this concept, "functions", and this link name, `[functions-approach]` (also used in other markdown files), is not appropriate. Every single approach has "functions" 😁 this name should be something distinctive about the approach. Maybe "multiple-clause-function", or shorter, "multiple-clauses"

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/introduction.md:68]: Like @Tuxified mentioned, the most idiomatic way should be presented here:
```suggestion
    true -> false
```

File: exercises/practice/leap/.approaches/functions/content.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/functions/content.md:1]:
```suggestion
# Multiple clause function
```

File: exercises/practice/leap/.approaches/flow/content.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/flow/content.md:1]: They're called "control flow structures" in the concept documents, so let's be consistent:
```suggestion
# Control flow structures
```

File: exercises/practice/leap/.approaches/flow/content.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/flow/content.md:18]:
```suggestion
Elixir provides for [control flow structures][hexdocs-structures]: `case`, `cond`, `if`, and `unless`. 
```

**Comment by angelikatyborska** [comment]
> What do you think about the level of detail? Too little? Too much?

I have no idea! What do other tracks do?

> How about external links?

Are you asking if it's too many or not enough? Or are you asking about their quality?

> Some of the formatting is difficult to see and evaluate unless the deep dive page is visible on Exercism. That's something I have learnt in other tracks. I think we should assume that despite doing our best to get it right, there might need to be a subsequent PR with a few corrections (e.g., headers, spacing around code examples, snippet formatting).

Agreed

**Comment by michalporeba** [comment]
> > What do you think about the level of detail? Too little? Too much?
> 
> I have no idea! What do other tracks do?
> 
The approach to approaches the tracks take are [not very consistent](https://forum.exercism.org/t/lack-of-convention-about-dig-deeper/2151). In that sense, the approaches are similar to concepts and exercise introductions. Some tracks are very detailed to the point of becoming a language tutorial. Other tracks are much more dry. Concepts can be lengthy and detailed, but they can only be more or less code examples. 

Personally, I like the approaches, because they offer more than what can be seen in the community solutions. They offer an opportunity to explain language features in context and how they apply to a problem. You can help somebody expand their knowledge and understanding. 

I hope they can help people like siraj-samsudeen, who [wrote on the forum](https://forum.exercism.org/t/feedback-on-elixir-track-after-doing-5-exercises-for-functional-feb/4195):
> When I looked at the community solutions after submitting my solution, the top-rated solutions were greek-and-latin to me - i could NOT make out much of what they were doing. There was no explanation in the code.

What I'm trying to do when I decide what to include and what not to do is go back to [this explanation of the idea by Jeremy](https://exercism.org/blog/dig-deeper). 

But, as I said, each track is different. As you are leading the Elixir one, I wonder what you think about it. If you don't normally look at approaches, the question might be more what you think about what I'm proposing here. If it works, I'll try to stay consistent from now on. 

> > How about external links?
> 
> Are you asking if it's too many or not enough? Or are you asking about their quality?
> 
Yes, both. I'm trying to be informative but not to explain everything that can be linked. I am also trying to limit the links to exercism and hexdocs.

File: exercises/practice/leap/.approaches/operators/content.md
Comment (michalporeba) [reply exercises/practice/leap/.approaches/operators/content.md:17]: Here are the reasons why I think it is still relevant, even if all inputs are valid, all the checks are pure functions. 

1. In the approach I'm showing different variations of the code. The option with the custom `divides?/2`  function is not necessarily guaranteed to be pure. 
2. The approach at the end I'm showing an option to capture the results in variables. Perhaps I should make it more explicit, but in this case, all checks are performed because the short-circuiting applies only to checking the value of the variables. 
3. In every exercise instructions, at the bottom, there instructions on how to use `IO.inspect/1` for debugging your solutions. It is possible, that somebody less experienced will be surprised that inspect added to a custom function, or perhaps directly to one of the checks. Consider this code

```elixir
rem(year, 4) == 0 and not IO.inspect(rem(year, 100) == 0) or rem(year, 400) == 0
```
When `rem(year, 4) != 0` the `IO.inspect/1` will not show anything because it will not be evaluated.

File: exercises/practice/leap/.approaches/introduction.md
Comment (michalporeba) [reply exercises/practice/leap/.approaches/introduction.md]: @angelikatyborska, how about `clauses` or `function-clauses` for the slug?

**Comment by michalporeba** [comment]
@angelikatyborska, it's been a long thread. Thanks for all the feedback. To help finalise this PR I think we are down to three things to decide on / adjust. 
1. Better slug for `functions`. My preferences are in this order: `clauses`, `function-clauses`, `multiple-function-clauses`. I favour shorter slugs to match the other two: `operators` and `flow`. But I'd be happy to move to longer for all of them. Perhaps: `control-flow`, `function-clauses`, `boolean-operators`. 
2. using the `non-negative` word. I think this is more accurate, but I would like to change the `number` to `divisor` there to align with the mathematical terms of the module operation.
3. short-circuiting. I'd like to keep it. A longer justification is available above. 
4. variations on `case`. I'd like to keep it with the adjusted wording to explain that `case on a tuple` is the most idiomatic approach. 

I'd like to finish this before the end of the weekend if you are available. I'll be available from time to time until about 20:00 today (your time). Tomorrow I will be available much less, mostly in the evening.

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [reply exercises/practice/leap/.approaches/introduction.md]: sounds good too

File: exercises/practice/leap/.approaches/operators/content.md
Comment (angelikatyborska) [reply exercises/practice/leap/.approaches/operators/content.md:17]: ok, I'm convinced 🙂

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [reply exercises/practice/leap/.approaches/introduction.md:12]: alright, no problem, we can keep it

**Comment by angelikatyborska** [comment]
> As you are leading the Elixir one, I wonder what you think about it. If you don't normally look at approaches, the question might be more what you think about what I'm proposing here. If it works, I'll try to stay consistent from now on.

If you want to take the detailed approach, I'm perfectly fine with that. I will just make sure that all the information in there is correct and consistent with the other docs 🙂 

> To help finalise this PR I...

To answer shortly:

> 1. Better slug for functions

I'm fine with keeping it short as long it's accurate, "clauses" is short and accurate, so go for that!

> 2. using the non-negative word

It was just a nitpick, not a problem, can stay!

> 3. short-circuiting

also can stay

> 4. variations on case. I'd like to keep it with the adjusted wording to explain that case on a tuple is the most idiomatic approach.

alright

I'll try to double-check the links later today and I'll make sure that the PR gets merged until the end of the weekend 👍

**Review by angelikatyborska — APPROVED** [review]
I pushed two tiny commits with some reformatting and grammar fixes so that it's faster than leaving comments about them.

I have no thoughts about the links 😅 looks fine to me.

Let's merge and see how it looks on the website, maybe people will have some feedback.

I'm approving and you can press the merge button when you're ready 🚀

**Comment by angelikatyborska** [comment]
Hmm, or maybe not, let's see those CI failures 🙃

**Comment by angelikatyborska** [comment]
Right, the link checker is going to fail because there are links to the new content that's not online yet:

```

## Errors per input

### Errors in exercises/practice/leap/.approaches/clauses/content.md

* [404] [https://exercism.org/tracks/elixir/exercises/leap/approaches/operators](https://exercism.org/tracks/elixir/exercises/leap/approaches/operators) | Failed: Network error: Not Found

### Errors in exercises/practice/leap/.approaches/flow/content.md

* [404] [https://exercism.org/tracks/elixir/exercises/leap/approaches/clauses](https://exercism.org/tracks/elixir/exercises/leap/approaches/clauses) | Failed: Network error: Not Found

### Errors in exercises/practice/leap/.approaches/introduction.md

* [404] [https://exercism.org/tracks/elixir/exercises/leap/approaches/clauses](https://exercism.org/tracks/elixir/exercises/leap/approaches/clauses) | Failed: Network error: Not Found
* [404] [https://exercism.org/tracks/elixir/exercises/leap/approaches/cond](https://exercism.org/tracks/elixir/exercises/leap/approaches/cond) | Failed: Network error: Not Found
* [404] [https://exercism.org/tracks/elixir/exercises/leap/approaches/operators](https://exercism.org/tracks/elixir/exercises/leap/approaches/operators) | Failed: Network error: Not Found
 ```

We can ignore that and merge anyway

**Comment by michalporeba** [comment]
@angelikatyborska I cannot merge, you'll have to do the button pushing. 
image
But let's do it and see how it looks, get some feedback.

**Comment by angelikatyborska** [comment]
Ah shit, sorry, I keep forgetting how it works with permissions 😬

**Comment by angelikatyborska** [comment]
It is live: https://exercism.org/tracks/elixir/exercises/leap/dig_deeper

thanks again!

File: exercises/practice/leap/.approaches/introduction.md
Comment (angelikatyborska) [inline exercises/practice/leap/.approaches/introduction.md:81]: This one broken link snuck in to `main`. Could you open a PR with a fix? If you do it, I can just merge, but if I do it, I will need to wait for another person with write rights to review because I can't approve my own PRs 😇

**Pull request merged**
